cppcache/src/PdxRemoteReader.cpp

```diff
@@ -34,571 +34,391 @@ PdxRemoteReader::~PdxRemoteReader() {
-      PdxLocalReader::resettoPdxHead();
-      m_dataInput->advanceCursor(position);
-      auto retVal = PdxLocalReader::readObjectArray(fieldName);
-      PdxLocalReader::resettoPdxHead();
-      return retVal;
-    }
+  int position = m_pdxType->getFieldPosition(fieldName, m_offsetsBuffer,
+                                             m_offsetSize, m_serializedLength);
+
+  if (position == -1) {
+    return {};
   }
+
+  PdxLocalReader::resettoPdxHead();
+  m_dataInput->advanceCursor(position);
+
+  auto result = PdxLocalReader::readObjectArray(fieldName);
+  PdxLocalReader::resettoPdxHead();
+  return result;
 }
 
 int8_t** PdxRemoteReader::readArrayOfByteArrays(const std::string& fieldName,
                                                 int32_t& arrayLength,
                                                 int32_t** elementLength) {
-  int choice = m_localToRemoteMap[m_currentIndex++];
-
-  switch (choice) {
-    case -2:
-      return PdxLocalReader::readArrayOfByteArrays(
-          fieldName, arrayLength, elementLength);  // in same order
-    case -1:
-      return nullptr;  // null value
-    default: {
-      // sequence id read field and then update
-      int position = m_pdxType->getFieldPosition(
-          choice, m_offsetsBuffer, m_offsetSize, m_serializedLength);
-      PdxLocalReader::resettoPdxHead();
-      m_dataInput->advanceCursor(position);
-      int8_t** retVal = PdxLocalReader::readArrayOfByteArrays(
-          fieldName, arrayLength, elementLength);
-      PdxLocalReader::resettoPdxHead();
-      return retVal;
-    }
+  int position = m_pdxType->getFieldPosition(fieldName, m_offsetsBuffer,
+                                             m_offsetSize, m_serializedLength);
+
+  if (position == -1) {
+    return nullptr;
   }
+
+  PdxLocalReader::resettoPdxHead();
+  m_dataInput->advanceCursor(position);
+
+  auto result = PdxLocalReader::readArrayOfByteArrays(fieldName, arrayLength,
+                                                      elementLength);
+
+  PdxLocalReader::resettoPdxHead();
+  return result;
 }
 
 std::shared_ptr<CacheableDate> PdxRemoteReader::readDate(
     const std::string& fieldName) {
-  int choice = m_localToRemoteMap[m_currentIndex++];
-
-  switch (choice) {
-    case -2:
-      return PdxLocalReader::readDate(fieldName);  // in same order
-    case -1:
-      return nullptr;
-    default: {
-      // sequence id read field and then update
-      int position = m_pdxType->getFieldPosition(
-          choice, m_offsetsBuffer, m_offsetSize, m_serializedLength);
-      PdxLocalReader::resettoPdxHead();
-      m_dataInput->advanceCursor(position);
-      auto retVal = PdxLocalReader::readDate(fieldName);
-      PdxLocalReader::resettoPdxHead();
-      return retVal;
-    }
+  int position = m_pdxType->getFieldPosition(fieldName, m_offsetsBuffer,
+                                             m_offsetSize, m_serializedLength);
+
+  if (position == -1) {
```

Comment:
I wonder, if with all this refactoring, we should define a `constexpr` for this `-1` sentinel value. If you make further changes I would suggest you do this too, otherwise something to keep in your mind for later.